src/scr_flush_file.c

```diff
@@ -40,6 +45,167 @@
+/*
+ * Given a path to a prefix directory, the contents of the flush file,
+ * an id, and the path to the flush_file, generate the summary.scr file and
+ * update flush.scr to show that we're no longer flushing.
+ *
+ * Returns 0 on success.  Returns 1 if there's no corresponding dataset_id
+ * in the flush file.
+ */
+int write_summary_file(char* prefix, kvtree* flush_file, int dataset_id,
+  char* flush_file_path)
+{
+
+  /* define metadata directory for dataset */
+  char* summary_file = NULL;
+  int rc;
+  kvtree* flush_key_dataset;
+  kvtree* summary;
+
+  flush_key_dataset = kvtree_get_kv_int(flush_file, SCR_FLUSH_KEY_DATASET, dataset_id);
+  if (flush_key_dataset == NULL) {
+    printf("%s: no flush file entry for dataset %d\n", __func__, dataset_id);
+    return 1; /* No entry for this dataset */
+  }
+
+  summary = kvtree_get(flush_key_dataset, SCR_FLUSH_KEY_DSETDESC);
+
+  asprintf(&summary_file, "%s/.scr/scr.dataset.%d/summary.scr", prefix, dataset_id);
+
+  rc = scr_flush_summary_file(summary, 1, summary_file);
+
+  scr_free(&summary_file);
+
+  /* Update the flush file that we're not longer transferring */
+  scr_flush_file_location_unset_with_path(dataset_id,
+    SCR_FLUSH_KEY_LOCATION_SYNC_FLUSHING, flush_file_path);
+
+  return rc;
+}
+
+/*
+ * Given a path to a state_file, resume and finalize all transfers for all
+ * files in the state_file.
+ */
+int resume_transfer(char* state_file_path)
+{
+  int id;
+  int rc;
+  AXL_Init();
+
+  id = AXL_Create(AXL_XFER_STATE_FILE, "scr", state_file_path);
+  if (id < 0) {
+    scr_err("Error: AXL_Create() = %d\n", id);
+    rc = id;
+    goto end;
+  }
+
+  rc = AXL_Resume(id);
+  if (rc != AXL_SUCCESS) {
+    scr_err("Error: AXL_Resume(%d) = %d\n", id, rc);
+    goto end;
+  }
+
+  rc = AXL_Wait(id);
+  if (rc != AXL_SUCCESS) {
+    scr_err("Error: AXL_Wait(%d), = %d\n", id, rc);
+    goto end;
+  }
+
+  rc = AXL_Free(id);
+  if (rc != AXL_SUCCESS) {
+    scr_err("Error: AXL_Free(%d), = %d\n", id, rc);
+    goto end;
+  }
+
+  rc = AXL_Finalize();
+  if (rc != AXL_SUCCESS) {
+    scr_err("Error: AXL_Finalize(), = %d\n", id, rc);
+  }
+end:
+  return rc;
+}
+
+/*
+ * Resume and wait for any previous transfers to complete, and finalize them.
+ *
+ * This only resumes/waits for the AXL transfers to complete.  It does not
+ * update SCR's flush file nor write the summary file.
+ *
+ * Returns 0 on success, non-zero otherwise.
+ */
+int resume_transfers(char* prefix, int dataset_id)
+{
+  char* rank2file_path = NULL;
+  int rc;
+  int ranks;
+  char num_str[21]; /* Max number in uint64_t is 20 digits + '\0' */
+  unsigned long i;
+  char* file;
+  char* state_file;
+  kvtree* ranks_tree;
+  kvtree* rank_subtree;
+
+  ranks_tree = kvtree_new();
+
+  asprintf(&rank2file_path, "%s/.scr/scr.dataset.%d/rank2file", prefix, dataset_id);
+
+  rc = kvtree_read_scatter_single(rank2file_path, ranks_tree);
+  if (rc) {
+    scr_err("kvtree_read_scatter_single(%s) = %d\n", __func__, rank2file_path, rc);
```

Comment:
I don't think we need ``\n`` on scr_err calls.  If I'm remembering correctly, let's remove those.
```suggestion
    scr_err("kvtree_read_scatter_single(%s) = %d", __func__, rank2file_path, rc);
```

Reply:
For scr_err messages, let's also update those to end with ``@ %s:%d`` and print __FILE__ and __LINE__ values like this example:
```
      scr_err("Failed to read SCR_FLUSH_ASYNC_BW successfully @ %s:%d",
        __FILE__, __LINE__
      );
```

Reply:
And let's double check each of those that the printf format string matches the type and order of the arguments.  For example, I think this particular line is missing a ``%s`` somewhere or we should drop ``__func__``.
```
scr_err("kvtree_read_scatter_single(%s) = %d\n", __func__, rank2file_path, rc);
```
Those lines need extra manual care, since the compiler often can't catch them, and they don't execute at runtime since they are error messages.

Reply:
That looks like a debug print that snuck in, I'll remove it.  I'll change the other ones I added to include __FILE__ and __LINE__ if they don't have them already.